<!DOCTYPE html>
<html>
<head>
<title>protocol_for_assigning_taxonomy_to_cultures</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<style type="text/css">
/* GitHub stylesheet for MarkdownPad (http://markdownpad.com) */
/* Author: Nicolas Hery - http://nicolashery.com */
/* Version: b13fe65ca28d2e568c6ed5d7f06581183df8f2ff */
/* Source: https://github.com/nicolahery/markdownpad-github */

/* RESET
=============================================================================*/

html, body, div, span, applet, object, iframe, h1, h2, h3, h4, h5, h6, p, blockquote, pre, a, abbr, acronym, address, big, cite, code, del, dfn, em, img, ins, kbd, q, s, samp, small, strike, strong, sub, sup, tt, var, b, u, i, center, dl, dt, dd, ol, ul, li, fieldset, form, label, legend, table, caption, tbody, tfoot, thead, tr, th, td, article, aside, canvas, details, embed, figure, figcaption, footer, header, hgroup, menu, nav, output, ruby, section, summary, time, mark, audio, video {
  margin: 0;
  padding: 0;
  border: 0;
}

/* BODY
=============================================================================*/

body {
  font-family: Helvetica, arial, freesans, clean, sans-serif;
  font-size: 14px;
  line-height: 1.6;
  color: #333;
  background-color: #fff;
  padding: 20px;
  max-width: 960px;
  margin: 0 auto;
}

body>*:first-child {
  margin-top: 0 !important;
}

body>*:last-child {
  margin-bottom: 0 !important;
}

/* BLOCKS
=============================================================================*/

p, blockquote, ul, ol, dl, table, pre {
  margin: 15px 0;
}

/* HEADERS
=============================================================================*/

h1, h2, h3, h4, h5, h6 {
  margin: 20px 0 10px;
  padding: 0;
  font-weight: bold;
  -webkit-font-smoothing: antialiased;
}

h1 tt, h1 code, h2 tt, h2 code, h3 tt, h3 code, h4 tt, h4 code, h5 tt, h5 code, h6 tt, h6 code {
  font-size: inherit;
}

h1 {
  font-size: 28px;
  color: #000;
}

h2 {
  font-size: 24px;
  border-bottom: 1px solid #ccc;
  color: #000;
}

h3 {
  font-size: 18px;
}

h4 {
  font-size: 16px;
}

h5 {
  font-size: 14px;
}

h6 {
  color: #777;
  font-size: 14px;
}

body>h2:first-child, body>h1:first-child, body>h1:first-child+h2, body>h3:first-child, body>h4:first-child, body>h5:first-child, body>h6:first-child {
  margin-top: 0;
  padding-top: 0;
}

a:first-child h1, a:first-child h2, a:first-child h3, a:first-child h4, a:first-child h5, a:first-child h6 {
  margin-top: 0;
  padding-top: 0;
}

h1+p, h2+p, h3+p, h4+p, h5+p, h6+p {
  margin-top: 10px;
}

/* LINKS
=============================================================================*/

a {
  color: #4183C4;
  text-decoration: none;
}

a:hover {
  text-decoration: underline;
}

/* LISTS
=============================================================================*/

ul, ol {
  padding-left: 30px;
}

ul li > :first-child, 
ol li > :first-child, 
ul li ul:first-of-type, 
ol li ol:first-of-type, 
ul li ol:first-of-type, 
ol li ul:first-of-type {
  margin-top: 0px;
}

ul ul, ul ol, ol ol, ol ul {
  margin-bottom: 0;
}

dl {
  padding: 0;
}

dl dt {
  font-size: 14px;
  font-weight: bold;
  font-style: italic;
  padding: 0;
  margin: 15px 0 5px;
}

dl dt:first-child {
  padding: 0;
}

dl dt>:first-child {
  margin-top: 0px;
}

dl dt>:last-child {
  margin-bottom: 0px;
}

dl dd {
  margin: 0 0 15px;
  padding: 0 15px;
}

dl dd>:first-child {
  margin-top: 0px;
}

dl dd>:last-child {
  margin-bottom: 0px;
}

/* CODE
=============================================================================*/

pre, code, tt {
  font-size: 12px;
  font-family: Consolas, "Liberation Mono", Courier, monospace;
}

code, tt {
  margin: 0 0px;
  padding: 0px 0px;
  white-space: nowrap;
  border: 1px solid #eaeaea;
  background-color: #f8f8f8;
  border-radius: 3px;
}

pre>code {
  margin: 0;
  padding: 0;
  white-space: pre;
  border: none;
  background: transparent;
}

pre {
  background-color: #f8f8f8;
  border: 1px solid #ccc;
  font-size: 13px;
  line-height: 19px;
  overflow: auto;
  padding: 6px 10px;
  border-radius: 3px;
}

pre code, pre tt {
  background-color: transparent;
  border: none;
}

kbd {
    -moz-border-bottom-colors: none;
    -moz-border-left-colors: none;
    -moz-border-right-colors: none;
    -moz-border-top-colors: none;
    background-color: #DDDDDD;
    background-image: linear-gradient(#F1F1F1, #DDDDDD);
    background-repeat: repeat-x;
    border-color: #DDDDDD #CCCCCC #CCCCCC #DDDDDD;
    border-image: none;
    border-radius: 2px 2px 2px 2px;
    border-style: solid;
    border-width: 1px;
    font-family: "Helvetica Neue",Helvetica,Arial,sans-serif;
    line-height: 10px;
    padding: 1px 4px;
}

/* QUOTES
=============================================================================*/

blockquote {
  border-left: 4px solid #DDD;
  padding: 0 15px;
  color: #777;
}

blockquote>:first-child {
  margin-top: 0px;
}

blockquote>:last-child {
  margin-bottom: 0px;
}

/* HORIZONTAL RULES
=============================================================================*/

hr {
  clear: both;
  margin: 15px 0;
  height: 0px;
  overflow: hidden;
  border: none;
  background: transparent;
  border-bottom: 4px solid #ddd;
  padding: 0;
}

/* TABLES
=============================================================================*/

table th {
  font-weight: bold;
}

table th, table td {
  border: 1px solid #ccc;
  padding: 6px 13px;
}

table tr {
  border-top: 1px solid #ccc;
  background-color: #fff;
}

table tr:nth-child(2n) {
  background-color: #f8f8f8;
}

/* IMAGES
=============================================================================*/

img {
  max-width: 100%
}
</style>
</head>
<body>
<h2>Protocol for assigning taxonomy to cultures</h2>
<h3>Van Bael Lab</h3>
<p>by Steve Formel</p>
<p>April 19, 2018</p>
<hr />
<h3>Notes</h3>
<ol>
<li>
<p>This is a more reproducible method than what we were doing previously.  But hopefully someone comes up with something better than this in the future.</p>
</li>
<li>
<p>Probably the easiest thing to go wrong with this protocol is using incorrect filepaths.  Make sure you have replaced my Cypress folder name, steve, with your folder in the filepaths.</p>
</li>
<li>
<p>Up to this point you should have:</p>
<pre><code>a. Cultured bacteria or fungi
b. Extracted DNA
c. Sanger sequenced the sample
d. Used a program like Mesquite to re-call the bases as PHRED bases
f. Used a program like Sequencher to edit the sequences.
        i. Edited the forward and reverse read into one 
   consensus FASTA sequence
        ii. although you could do the following with just the forward or reverse read
</code></pre>

</li>
</ol>
<h3>Make multifasta sequence</h3>
<ol>
<li>
<p>Upload all of your sequences to one folder in Cypress via Filezilla.</p>
<p>a. If you don't have a Cypress account, Sunshine can help you get one.   <a href="https://wiki.hpc.tulane.edu/trac/wiki/cypress/about">Link to Cypress Instructions</a></p>
</li>
<li>
<p>Log in to Cypress using terminal on a mac or putty on a Windows machine.</p>
</li>
<li>
<p>Navigate to your folder with the sequences.</p>
<p>a. If you're not certain how to do this, google &quot;cd unix&quot;.</p>
</li>
<li>
<p>Concatenate all your sequences into one multi-fasta file</p>
<pre><code>cat *.fasta &gt; all_my_seqs.fasta
</code></pre>

</li>
</ol>
<h3>Clustering: Optional</h3>
<p>People in our lab have done analysis that was clustered, assigned taxonomy to representatives from the clusters and then statistically analyzed.  Other projects in our lab have assigned taxonomy to every isolate from their project and then statistically analyzed the results using taxonomy as &quot;clusters&quot;.  I think you can argue for both, but I prefer to cluster first.</p>
<h4>De-novo clustering in QIIME</h4>
<p>If you have many isolates, you may want to cluster them at 97% in QIIME 1 or Sequencher.  Here are directions in QIIME 1.  Directions for Sequencher are elsewhere on the Van Bael Google Drive.  I used QIIME because it is a well-cited method (meant for 454, Illumina, and Sanger sequencing) and easily outputs your results.  If you choose to use Sequencher you need to jump through some hoops to get your clusters into a usable format.</p>
<p>This is a SLURM script to run on Cypress.  In the Cypress terminal, type:</p>
<pre><code>nano pick_denovo_otus.sh
</code></pre>

<p>And paste this into the editor:</p>
<pre><code>#!/bin/bash

#SBATCH --job-name=denovo_otus
#SBATCH --output=denovo_otus.output
#SBATCH --error=dneovo_otus.error
#SBATCH --qos=normal
#SBATCH --time=0-24:00:00
#SBATCH --nodes=1
#SBATCH --mem=64000
#SBATCH --mail-type=ALL

module load qiime/1.9.1

pick_de_novo_otus.py -i /lustre/project/svanbael/steve/all_my_seqs.fasta -o /lustre/project/svanbael/steve/denovo_otus -a -O 20
</code></pre>

<p>To exit:</p>
<pre><code>Press the control and x keys    
</code></pre>

<p>It will ask you if you want to save it.  
</p>
<pre><code>Hit y and then enter.  
</code></pre>

<p>You should be back in the regular terminal.  To run the job type:</p>
<pre><code>sbatch pick_denovo_otus.sh
</code></pre>

<p>To make sure your job is running type:</p>
<pre><code>squeue -u yourusername
</code></pre>

<p>You should see something like:</p>
<pre><code>JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
708647      defq make_ncb  sformel  R       0:06      1 cypress01-059
</code></pre>

<p>If you don't see anything, it didn't work for some reason.  The database will take 30-60 min to download and build.  To check that the job has finished type the above code again.  If the job has finished you should see:</p>
<pre><code>JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
</code></pre>

<h3>Make local ncbi nucleotide database.</h3>
<p>This is a slurm script.  You will run it through Cypress.</p>
<p>In the Cypress terminal type:</p>
<pre><code>nano make_ncbi_db.sh
</code></pre>

<p>Copy this code and paste it into the terminal.</p>
<pre><code>#!/bin/bash

#SBATCH --job-name=make_NCBI_db
#SBATCH --output=make_NCBI_db.output
#SBATCH --error=make_NCBI_db.error
#SBATCH --qos=normal
#SBATCH --time=0-24:00:00
#SBATCH --nodes=1
#SBATCH --mem=64000
#SBATCH --mail-type=ALL

##make directory and download and unzip files

##taken from: https://github.com/Joseph7e/Assign-Taxonomy-with-BLAST &amp; https://www.linuxquestions.org/questions/linux-newbie-8/how-to-untar-all-tar-files-in-a-directory-98963/

mkdir ncbi_nt_database
cd ncbi_nt_database

wget 'ftp://ftp.ncbi.nlm.nih.gov/blast/db/nt*.gz'

## Download NCBI's taxonomic data and GI (GenBank ID) taxonomic
## assignation.

wget 'ftp://ftp.ncbi.nlm.nih.gov/blast/db/taxdb.tar.gz'

for i in *.tar.gz; do echo working on $i; tar xvzf $i ; done

exit 0
</code></pre>

<p>To exit:</p>
<pre><code>Press the control and x keys    
</code></pre>

<p>It will ask you if you want to save it.  
</p>
<pre><code>Hit y and then enter.  
</code></pre>

<p>You should be back in the regular terminal. </p>
<p>To run the job type:</p>
<pre><code>sbatch make_ncbi_db.sh
</code></pre>

<h3>Run BLAST</h3>
<p>Make another script to BLAST your sequences.</p>
<p>Make sure that the filepaths in the script match where your ncbi database has been built.  <em>It is also important that your multifasta file is in the same folder as the ncbi database!</em> You can use Filezilla to move/copy files, or use the Unix command &quot;cp&quot;.  
</p>
<pre><code>#!/bin/bash

#SBATCH --job-name=my_BLAST
#SBATCH --output=my_BLAST.output
#SBATCH --error=my_BLAST.error
#SBATCH --qos=normal
#SBATCH --time=0-24:00:00
#SBATCH --nodes=1
#SBATCH --mem=64000
#SBATCH --mail-type=ALL

##Local blast on Cypress for culture based sequences

module load ncbi-blast/2.5.0+

blastn -db /lustre/project/svanbael/steve/ncbi_nt_database/nt -query /lustre/project/svanbael/steve/ncbi_nt_database/all_my_seqs.fasta  -num_threads 10 -out /lustre/project/svanbael/steve/my_blast_results.txt
</code></pre>

<h3>Run BLAST with output from QIIME denovo OTU picking</h3>
<p>If you picked your OTUs in QIIME with the above script, you need to copy the representative sequences of the OTUs to the folder with the ncbi database in it.</p>
<pre><code>cp /lustre/project/svanbael/steve/denovo_otus/rep_set/all_my_seqs_rep_set.fasta /lustre/project/svanbael/steve/ncbi_nt_database/
</code></pre>

<p>Once this is done, use the same script as above to do the BLAST, making sure that filepaths are correct.</p>
<h3>Filter BLAST results in MEGAN</h3>
<p>Use Filezilla to transfer your blast results (in the exmaple above, nameed my_blast_results.txt) to your computer.  
</p>
<p>I have installed MEGAN 6 community edition on one of our computers, but you can also install it on your own computer pretty easily.</p>
<p><a href="https://ab.inf.uni-tuebingen.de/software/megan6/download" title="Download MEGAN">Download MEGAN</a></p>
<p><em>Note that different version of MEGAN use different LCA algorithms!</em></p>
<h4>Import BLAST Results</h4>
<ol>
<li>In MEGAN click on File &gt; Import from BLAST</li>
<li>Click the folder icon on the right side of the top box.  Find the BLAST results file.</li>
<li>It should autodetect that this is a BLASTText file (Format pull down menu) and BlastN format (Mode pulldown menu)</li>
<li>Click &quot;Apply&quot; in the bottom right corner.</li>
<li>It will taker a while to process the info.</li>
</ol>
<h4>Fine-tuning taxonomy results</h4>
<p>The MEGAN results are based of Lowest Common Ancestry (LCA) of some subgroup of the BLAST results.  You refine these by clicking on:</p>
<p>Options &gt; LCA parameters</p>
<p>For the purposes of the algorithm Sanger sequencing is considered a short read, and the weightedLCA algorithm is the one the makers of MEGAN recommend at the time this was written.  The easiest solution is probably to use the default parameters (in terms of writing up your methods) but feel free to do some more reading to fine-tune your results.</p>
<h4>Export Results to Taxonomy Table</h4>
<ol>
<li>Press Control and a to select all the nodes on the tree.</li>
<li>Click File &gt; Export &gt; Text (CSV) Format</li>
<li>Change the top pulldown menu to &quot;readName_to_taxonPath&quot;</li>
<li>Make sure the middle pulldown menu is &quot;assigned&quot;</li>
<li>Click OK and select where you want the file to be saved.</li>
<li>Open this file in excel.  If you're not sure how to separate the taxonomy into individual columns, google &quot;Excel text to columns&quot;.</li>
</ol>
<h4>Maintaining Records</h4>
<p>Make sure you note what day you downloaded the ncbi database for your records.  It changes on a daily basis.</p>
<p>Make sure you keep your MEGAN project as part of your data.  It's the only way to see the Accession numbers of the BLAST hits that contribute to the taxonomic assignment of your sequence.  As far as I can tell, there is no easy way to export these numbers.  Furthermore, the number of hits that contributes to each taxonomic assignment varies (read about the LCA algorithm if you don't understand why) so it would be difficult to easily format the results into a table.</p>
<h4>Last thoughts</h4>
<p>The NCBI taxonomy is problematic, and different taxa and entries include different classifications, some of which are unranked.  Ultimately you will decide what taxonomy to include with your sequence when you upload it to genbank.  For simplicity in your analysis, you probably want to pare down your taxonomy to a traditional KPCOFGS taxonomy in excel.</p>

</body>
</html>
<!-- This document was created with MarkdownPad, the Markdown editor for Windows (http://markdownpad.com) -->
